### PR TITLE
Pass extra data (fe. accountIDs) to HTMLToMarkdown method & add possibility to disable certain rules

### PR DESCRIPTION
**Pull request by robertKozik**
This PR consists of two different changes
1.  `htmlToMarkdown` method now has additional, optional argument `extras` this argument adds the possibility to pass extra data which can be used in replacement function. For now is used to pass accountID to name and reportID to name maps
2. `replace` method has additional object key `disabledRules`. Every rule name from this array will be skipped in parsing.

### Fixed Issues
$ https://github.com/Expensify/App/issues/40480
https://github.com/Expensify/App/issues/39550

# Tests
1. What unit/integration tests cover your change? What autoQA tests cover your change?
The changes, which this PR introduces, allows to more precisely interact with parser. To be precise, now we will be able to pass application state to the parser itself. We wrote couple of unit tests which mocks passing `extras` object with account ID and reportID maps to ensure that parsing shortened room and user mentions tags works as intended.

1. What tests did you perform that validates your changed worked?
We also tested our changes inside E/App. It's not yet supported on main branch, but all changes worked as intended

# QA
1. What does QA need to do to validate your changes?
These changes cannot be tested on E/App yet. Only follow-up PR will be using these changes
1. What areas to they need to test for regressions?
As we were touching methods responsible for both HTML –> markdown and markdown -> HTML we should look for regressions in every rule and parsing.

### CODE REVIEW

**Comment by war-in** [comment]
I have read the CLA Document and I hereby sign the CLA

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -453,7 +457,31 @@ export default class ExpensiMark {
 
                     return `!(${g2})`;
                 }
-            }
+            },
+            {
+                name: 'roomMention',
```
Comment (war-in) [inline lib/ExpensiMark.js:462]: Now I wonder if we should keep the `room` here or change it to the `report`

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -768,10 +836,11 @@ export default class ExpensiMark {
      * Replaces HTML with markdown
      *
      * @param {String} htmlString
+     * @param {Object} extras
```
Comment (war-in) [inline lib/ExpensiMark.js:839]: `ExtrasObject`? 🤔

File: lib/ExpensiMark.js [hunk 3]
```diff
@@ -795,13 +867,16 @@ export default class ExpensiMark {
      * Convert HTML to text
      *
      * @param {String} htmlString
+     * @param {Object} extras
```
Comment (war-in) [inline lib/ExpensiMark.js:870]: Same here

File: lib/ExpensiMark.js [hunk 4]
```diff
@@ -795,13 +867,16 @@ export default class ExpensiMark {
...
+     *
      * @returns {String}
      */
-    htmlToText(htmlString) {
+    htmlToText(htmlString, extras) {
         let replacedText = htmlString;
 
         this.htmlToTextRules.forEach((rule) => {
-            replacedText = replacedText.replace(rule.regex, rule.replacement);
+            const replacementFunction = typeof rule.replacement === 'function' ? (...args) => rule.replacement(...args, extras) : rule.replacement;
```
Comment (war-in) [inline lib/ExpensiMark.js:878]: Should we add this comment here too?

```
// if replacement is a function, we want to pass optional extras to it
```

File: lib/ExpensiMark.js (hunk 1 above)
Comment (robertKozik) [reply lib/ExpensiMark.js]: change to report :P

File: lib/ExpensiMark.d.ts [hunk 5]
```diff
@@ -1,11 +1,13 @@
-declare type Replacement = (...args: string[]) => string;
+declare type Replacement = (...args: string[], extras?: ExtrasObject) => string;
 declare type Name =
     | 'codeFence'
     | 'inlineCodeBlock'
     | 'email'
     | 'link'
     | 'hereMentions'
+    | 'roomMentions'
```
Comment (rlinoz) [inline lib/ExpensiMark.d.ts:8]: I'm confused, why we need the `roomMentions` here if we renamed it o `reportMentions`?

**Review by rlinoz — COMMENTED** [review]
Looks good, just one question

File: lib/ExpensiMark.d.ts (hunk 5 above)
Comment (robertKozik) [reply lib/ExpensiMark.d.ts:8]: Thanks! Missed that while implementing

**Review by rlinoz — APPROVED** [review]
Thanks!

**Comment by rlinoz** [comment]
Merging to unblock other issues

**Pull request merged**